File: mcts.py
import math
import random

class MCTSNode:
    def __init__(self, state, parent=None, action=None):
        self.state = state
        self.parent = parent
        self.action = action
        self.children = []
        self.visits = 0
        self.wins = 0

    def is_fully_expanded(self):
        return len(self.children) == len(self.state.get_legal_actions())

    def best_child(self, exploration_param=math.sqrt(2)):
        best_score = -float('inf')
        best_child = None
        for child in self.children:
            exploit = child.wins / child.visits if child.visits > 0 else 0
            explore = math.sqrt(math.log(self.visits) / child.visits) if child.visits > 0 else float('inf')
            score = exploit + exploration_param * explore
            if score > best_score:
                best_score = score
                best_child = child
        return best_child

class MCTS:
    def __init__(self, simulations=1000):
        self.simulations = simulations

    def search(self, initial_state):
        root = MCTSNode(initial_state)

        for _ in range(self.simulations):
            node = root
            state = initial_state.clone()

            # Selection
            while not state.is_terminal() and node.is_fully_expanded():
                node = node.best_child()
                state = state.take_action(node.action)

            # Expansion
            if not state.is_terminal():
                legal_actions = state.get_legal_actions()
                tried_actions = [child.action for child in node.children]
                untried_actions = [a for a in legal_actions if a not in tried_actions]

                action = random.choice(untried_actions)
                new_state = state.take_action(action)
                child_node = MCTSNode(new_state, parent=node, action=action)
                node.children.append(child_node)
                node = child_node
                state = new_state

            # Simulation
            while not state.is_terminal():
                action = random.choice(state.get_legal_actions())
                state = state.take_action(action)

            # Backpropagation
            result = state.get_result()
            while node is not None:
                node.visits += 1
                node.wins += result
                node = node.parent

        return max(root.children, key=lambda c: c.visits).action

File: src/main/java/mcts/tictactoe/MCTS.java
package mcts.tictactoe;

import mcts.core.Move;
import mcts.core.Node;
import mcts.core.State;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * Class to represent a Monte Carlo Tree Search for TicTacToe.
 */
public class MCTS {

    public static void main(String[] args) {
        TicTacToe game = new TicTacToe();
        TicTacToeNode rootNode = new TicTacToeNode(game.new TicTacToeState());
        MCTS mcts = new MCTS(rootNode, Math.sqrt(2));
        mcts.runSearch(100_000);
        TicTacToe.TicTacToeMove move = mcts.bestMove();
        System.out.printf("Best move: %d,%d by player %d\n", move.move()[0], move.move()[1], move.player());
    }

    /**
     * Run the MCTS algorithm for the given number of iterations.
     */
    public void runSearch(int iterations) {
        for (int i = 0; i < iterations; i++) {
            // 1. SELECTION
            List<Node<TicTacToe>> path = new ArrayList<>();
            Node<TicTacToe> node = root;
            path.add(node);
            while (!node.isLeaf() && !node.children().isEmpty()) {
                node = selectUCT(node);
                path.add(node);
            }

            // 2. EXPANSION
            if (!node.isLeaf() && node.children().isEmpty()) {
                node.explore();           // adds all immediate children
                // pick one child to simulate
                node = node.children().iterator().next();
                path.add(node);
            }

            // 3. SIMULATION (rollout)
            State<TicTacToe> state = node.state();
            int result = simulate(state);

            // 4. BACKPROPAGATION
            for (Node<TicTacToe> n : path) {
                TicTacToeNode tn = (TicTacToeNode)n;
                tn.recordPlayout();
                tn.recordWin(result);
            }
        }
    }

    /**
     * Select the child of 'node' with highest UCT value.
     */
    private Node<TicTacToe> selectUCT(Node<TicTacToe> node) {
        Node<TicTacToe> best = null;
        double bestValue = Double.NEGATIVE_INFINITY;
        double parentPlayouts = node.playouts();
        for (Node<TicTacToe> child : node.children()) {
            double childPlayouts = child.playouts();
            double uctValue;
            if (childPlayouts == 0) {
                uctValue = Double.POSITIVE_INFINITY;
            } else {
                double winRate = (double) child.wins() / childPlayouts;
                uctValue = winRate + Cp * Math.sqrt(Math.log(parentPlayouts) / childPlayouts);
            }
            if (uctValue > bestValue) {
                bestValue = uctValue;
                best = child;
            }
        }
        return best;
    }

    /**
     * Do a random playout from the given state to a terminal state.
     * @return the winner (0 or 1), or -1 for a draw.
     */
    private int simulate(State<TicTacToe> s) {
        State<TicTacToe> cur = s;
        while (!cur.isTerminal()) {
            int player = cur.player();
            Move<TicTacToe> m = cur.chooseMove(player);
            cur = cur.next(m);
        }
        Optional<Integer> winner = cur.winner();
        return winner.orElse(-1);
    }

    /**
     * After search, pick the most visited child as the best move.
     */
    public TicTacToe.TicTacToeMove bestMove() {
        State<TicTacToe> rootState = root.state();
        int player = rootState.player();

        // 1. pick the most‑visited child
        Node<TicTacToe> bestChild = null;
        int maxPlayouts = -1;
        for (Node<TicTacToe> c : root.children()) {
            if (c.playouts() > maxPlayouts) {
                maxPlayouts = c.playouts();
                bestChild = c;
            }
        }

        // 2. find which Move actually transitions from rootState → bestChild.state()
        for (Move<TicTacToe> m : rootState.moves(player)) {
            State<TicTacToe> s2 = rootState.next(m);
            if (s2.equals(bestChild.state())) {
                // this is the move you want
                return (TicTacToe.TicTacToeMove) m;
            }
        }

        throw new RuntimeException("bestMove: no matching move found");
    }


    public MCTS(Node<TicTacToe> root, double Cp) {
        this.root = root;
        this.Cp = Cp;
    }

    private final Node<TicTacToe> root;
    private final double Cp;
}